### PR TITLE
MM-45317: global drafts endpoints and ws events

### PR DESCRIPTION
**Pull request by koox00**
#### Summary

Adds endpoints:
- create/update drafts
- delete draft
- get drafts

Adds WS events:
- draft_updated
- draft_created
- draft_deleted

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-45317

#### Related Pull Requests
- https://github.com/mattermost/mattermost-webapp/pull/8869

#### Release Note

```release-note
Added support for synced drafts
```

### CODE REVIEW

**Review by esethna — APPROVED** [review]
Amazing! Great work @koox00 @matthewbirtch @vish9812

**Comment by agarciamontoro** [comment]
Please remember to add coverage for this new feature in the load-test tool :pray: Ping me if you need any help with that, I'm happy to jump on a call anytime :)

**Review by Willyfrog — APPROVED** [review]
LGTM

**Comment by esethna** [comment]
@jgilliam17 can you help start a QA pass on the draft PRs (server+web) given dev reviews are wrapping up?

**Comment by jgilliam17** [comment]
@koox00
I've tested upgrading the test server (with old drafts) to version with PR changes and haven't seen any loss of existing drafts.
I do see that Global drafts item is not automatically added for older, pre upgrade drafts. It is shown for newer, post upgrade, drafts as expected.
Let me know if this is something we can improve on. Thanks

**Comment by mylonsuren** [comment]
Load test grafana: http://18.212.200.148:3000

image


**Drafts:**
image

**Comment by mylonsuren** [comment]
> @koox00 I've tested upgrading the test server (with old drafts) to version with PR changes and haven't seen any loss of existing drafts. I do see that Global drafts item is not automatically added for older, pre upgrade drafts. It is shown for newer, post upgrade, drafts as expected. Let me know if this is something we can improve on. Thanks

Thanks @jgilliam17! This is expected behaviour, there is a ticket https://mattermost.atlassian.net/browse/MM-48433 to automatically add existing pre-upgrade drafts to the drafts page.

Just to confirm though, existing pre-upgrade drafts are visible in the channel/thread, and then appear in the global drafts page after interacting with them?

File: app/draft.go
Comment (vish9812) [inline app/draft.go:242]: Delete doesn't return NotFound error.

**Review by vish9812 — CHANGES_REQUESTED** [review]

File: store/sqlstore/draft_store.go
Comment (vish9812) [inline store/sqlstore/draft_store.go:142]: Is it a valid case that the `Message` can be empty in DB?

File: store/storetest/draft_store.go
Comment (mylonsuren) [reply store/storetest/draft_store.go:12]: @vish9812 Do we add test cases here? Looking at the other stores this seems to be the pattern.

File: store/sqlstore/draft_store.go
Comment (mylonsuren) [reply store/sqlstore/draft_store.go:142]: It is a valid case, as there could be drafts with just file attachments, but good catch as this was blocking those drafts from being fetched. Removing this line :+1:

File: app/draft.go
Comment (mylonsuren) [reply app/draft.go:242]: 👍

File: app/draft.go
Comment (mylonsuren) [reply app/draft.go:205]: Yeah that makes sense, updated the log :+1:

File: app/draft.go
Comment (mylonsuren) [reply app/draft.go:112]: Sounds good, updated the error handling for all store function calls

File: app/draft_test.go
Comment (mylonsuren) [reply app/draft_test.go:101]: 😅 Looks like an extra line from copy/paste.

File: store/sqlstore/draft_store.go
Comment (mylonsuren) [reply store/sqlstore/draft_store.go:59]: Added `"DeleteAt": 0` to prevent fetching of deleted drafts :+1:

File: store/sqlstore/draft_store.go
Comment (mylonsuren) [reply store/sqlstore/draft_store.go:106]: You're right, query has been updated to only update relevant fields and prevent updating deleted drafts.

File: app/draft.go
Comment (mylonsuren) [reply app/draft.go:20]: 👍

File: app/draft.go
Comment (mylonsuren) [reply app/draft.go:80]: Yes this line is correct, I've update the other methods to check both config and feature flag

File: store/storetest/draft_store.go
Comment (vish9812) [reply store/storetest/draft_store.go:12]: A small file for reference - https://github.com/mattermost/mattermost-server/blob/MM-46410_priority-badges/store/storetest/post_priority_store.go

**Review by vish9812 — APPROVED** [review]
All changes look good.
Great work guys 👍 
Just add the Store tests and it's good to go.

**Comment by jgilliam17** [comment]
> Just to confirm though, existing pre-upgrade drafts are visible in the channel/thread, and then appear in the global drafts page after interacting with them?

Yes, existing pre-upgrade drafts are visible in the channel/thread, and they do appear in the global drafts after I continue to edit the draft post 

Seeing one issue though, the newly interacted draft shows up in global drafts for both town square and off topic channels, not just the town square where it was originally added before the update.
Those are the only 2 channels this test user belong to, not sure if the post would duplicated to all channels if there were more
Screen Shot 2022-11-23 at 1 38 01 PM

**Comment by mylonsuren** [comment]
> > Just to confirm though, existing pre-upgrade drafts are visible in the channel/thread, and then appear in the global drafts page after interacting with them?
> 
> @mylonsuren Yes, existing pre-upgrade drafts are visible in the channel/thread, and they do appear in the global drafts after I continue to edit the draft post
> 
> Seeing one issue though, the newly interacted draft shows up in global drafts for both town square and off topic channels, not just the town square where it was originally added before the update. Those are the only 2 channels this test user belong to, not sure if the post would duplicated to all channels if there were more Screen Shot 2022-11-23 at 1 38 01 PM

@jgilliam17 What are the steps to reproduce this exactly?

I tested this locally by:
1. Creating new drafts on the pre-upgrade server
2. Upgrading the server
3. Refreshing the page when prompted 

Wasn't able to repro, am I missing any steps?

**Comment by jgilliam17** [comment]
@mylonsuren sorry about delay
1. Creating new drafts on the pre-upgrade server with 2 teams
2. Upgrading the server
3. Refreshing the page when prompted
4. Open existing draft and edit it, do not post
5. Open global drafts - observe draft duplicated in off topic channel 

I am recreating my test server to try to reproduce again.

**Comment by mylonsuren** [comment]
@jgilliam17 Just a heads up as well the feature flag is now off by default per https://github.com/mattermost/mattermost-webapp/pull/8869#issuecomment-1325622874

**Comment by jgilliam17** [comment]
Thanks both @koox00 and @mylonsuren 
I am no longer able to reproduce the issue with doubling of drafts
I'l keep an eye on this after merge
Also, I created a split for GlobalDrafts and can enable if anyone else needs to take a look on the test server.

two small issues we can tackle after merge in a follow up, I can file tickets @esethna thoughts?
1. The size of the draft item on global draft is smaller when rhs is open, it seems that we are accommodating for expanded rhs before it is expanded
Screen Shot 2022-11-23 at 4 58 50 PM

2. Buttons on send or delete draft modals don't have expected hover effect
Screen Shot 2022-11-23 at 5 04 09 PM

**Comment by jgilliam17** [comment]
e2e-test MM_ENV="MM_FEATUREFLAGS_GLOBALDRAFTS=true"

**Comment by esethna** [comment]
@jgilliam17 ya let's file tickets so we can unblock merge, thanks!

**Comment by mylonsuren** [comment]
> Thanks both @koox00 and @mylonsuren I am no longer able to reproduce the issue with doubling of drafts I'l keep an eye on this after merge Also, I created a split for GlobalDrafts and can enable if anyone else needs to take a look on the test server.
> 
> two small issues we can tackle after merge in a follow up, I can file tickets @esethna thoughts?
> 
> 1. The size of the draft item on global draft is smaller when rhs is open, it seems that we are accommodating for expanded rhs before it is expanded
> 
> Screen Shot 2022-11-23 at 4 58 50 PM
> 
> 2. Buttons on send or delete draft modals don't have expected hover effect
> 
> Screen Shot 2022-11-23 at 5 04 09 PM

Thanks @jgilliam17! For the first one with the odd display behaviour with the RHS open, I found the issue so can push it to the webapp PR now if that's ok.

**Comment by jgilliam17** [comment]
@mylonsuren sounds good please ping me here when ready

**Comment by mylonsuren** [comment]
@jgilliam17 Fix has been pushed to the webapp PR, you may need to trigger a new test server to see the changes. 

Also, if you haven't created a new ticket for the modal button issue I can create it as there are a couple other fixes I'd like to put in post-merge.

**Comment by jgilliam17** [comment]
E2E report has more failures than master https://mattermost-cypress-report.s3.amazonaws.com/265982-8b39859-onprem-ent-server-pr-20614-mm-ee-test:8b39859/mochawesome.html
Since this is behind feature flag I will file follow up tickets to address the failures after the merge.

**Comment by jgilliam17** [comment]
Ticket for button hover effect  https://mattermost.atlassian.net/browse/MM-48613

**Comment by mylonsuren** [comment]
Thanks @jgilliam17, I'm going to resolve the conflicts first and I'll ping you once it's ready to merge.

**Comment by jgilliam17** [comment]
@mylonsuren acknowledgments PR is merged (both webapp and server) and there maybe some conflicts with these changes Kyriakos mentioned

**Review by jgilliam17 — APPROVED** [review]
The span of draft items when RHS is open, has been fixed. 
Looks good to merge after conflicts are resolved.
Screen Shot 2022-11-23 at 7 56 39 PM
I completely forgot that I created new tests when we first started working on this feature. I can create a ticket to automate in cypress if possible 
Screen Shot 2022-11-23 at 8 00 16 PM

File: model/feature_flags.go
Comment (esethna) [inline model/feature_flags.go:114]: @mylonsuren can you help submit a PR to set this default true for v7.6?

Comment (esethna) [reply model/feature_flags.go:114]: I think it should be submitted against master and then @amyblais will cherry pick to v7.6

Comment (amyblais) [reply model/feature_flags.go:114]: Ticket: https://mattermost.atlassian.net/browse/MM-48622.

**Pull request merged**
